### PR TITLE
Define new Pre- and PostEscapeAnalysis passes

### PR DESCRIPTION
**Pull request by hzongaro**
This change defines new optimization passes, preEscapeAnalysis and postEscapeAnalysis, that are designed to help Escape Analysis (EA) do a better job under voluntary OSR.

The new `preEscapeAnalysis` optimization looks for calls to `OSRInductionHelper` and adds a fake `prepareForOSR` call that references all live auto symrefs and pending pushes.  The goal is that any object that ends up as a candidate for stack allocation that appears to be used by such a fake `prepareForOSR` call can be heapified by EA at that point.  That will be delivered in a subsequent change to EA.

After EA, the `postEscapeAnalysis` pass will remove the fake calls to `prepareForOSR`.

These changes are a proper subset of the changes that were previously pull requested and reviewed under pull request #5737.  The changes to EA that take advantage of the work that PreEscapeAnalysis performs require that pass to actually run.  In order to do that, PreEscapeAnslysis and PostEscapeAnalysis need to be added to the optimization strategies that run EA in OMR, and in order to do, the optimizations need to exist.  Hence, this pull request.

Submitted on behalf of Andrew Craik <ajcraik@ca.ibm.com>

Signed-off-by:  Henry Zongaro <zongaro@ca.ibm.com>

### CODE REVIEW

**Comment by hzongaro** [comment]
I'm marking this change a WIP while I run some testing on it, to ensure I haven't broken anything.

**Comment by hzongaro** [comment]
@vijaysun-omr Vijay, as noted above, Andrew pointed out that the EA changes that are part of pull request #5737 need the Pre- and PostEscapeAnalysis passes to be operational.  But that requires a change in OMR, which will be done under [OMR pull request #3842](https://github.com/eclipse/omr/pull/3842).  In order to break the cycle of dependencies, the definitions and implementations of PreEscapeAnalysis and PostEscapeAnalysis have been pulled out into this pull request.

As you previously reviewed these changes under pull request #5737, may I ask you to re-review this subset here?  There are no substantive changes.

File: runtime/compiler/optimizer/PostEscapeAnalysis.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+         {
+         // Remove any fake prepareForOSR calls that were added to OSR
+         // induction blocks by TR_PreEscapeAnalysis
+         if (itr->getNode()->getNumChildren() == 1
+             && itr->getNode()->getFirstChild()->getOpCodeValue() == TR::call
+             && itr->getNode()->getFirstChild()->getSymbolReference()->getReferenceNumber() == TR_prepareForOSR)
+            {
+            dumpOptDetails(comp(), " Removing fake call prepareForOSR n%dn from induction block_%d\n", itr->getNode()->getFirstChild()->getGlobalIndex(), block->getNumber());
+            itr = itr->getPrevTreeTop();
+            if (optimizer()->getUseDefInfo() != NULL)
```
Comment (vijaysun-omr) [inline runtime/compiler/optimizer/PostEscapeAnalysis.cpp]: Should value numbering also be invalidated if it is non null here ?

File: runtime/compiler/optimizer/PreEscapeAnalysis.cpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+         continue;
+
+      for (TR::TreeTop *itr = block->getEntry(), *end = block->getExit(); itr != end; itr = itr->getNextTreeTop())
+         {
+         if (itr->getNode()->getNumChildren() == 1
+             && itr->getNode()->getFirstChild()->getOpCodeValue() == TR::call
+             && itr->getNode()->getFirstChild()->getSymbolReference()->isOSRInductionHelper())
+            {
+            _loads->clear();
+            if (optimizer()->getUseDefInfo() != NULL)
```
Comment (vijaysun-omr) [inline runtime/compiler/optimizer/PreEscapeAnalysis.cpp]: Should value numbering be invalidated here as well if it's non null ?

File: runtime/compiler/optimizer/PostEscapeAnalysis.cpp (hunk 1 above)
Comment (hzongaro) [reply runtime/compiler/optimizer/PostEscapeAnalysis.cpp]: Yes, it should.  Thank you for catching that.

File: runtime/compiler/optimizer/PreEscapeAnalysis.cpp (hunk 2 above)
Comment (hzongaro) [reply runtime/compiler/optimizer/PreEscapeAnalysis.cpp]: Yes, it should.  Thank you for catching that.

Comment (hzongaro) [reply runtime/compiler/optimizer/PreEscapeAnalysis.cpp]: Fixed in commit https://github.com/eclipse/openj9/pull/6654/commits/7f10d28781bc910fa1fcf09c5a429a2be1e55cf1

File: runtime/compiler/optimizer/PostEscapeAnalysis.cpp (hunk 1 above)
Comment (hzongaro) [reply runtime/compiler/optimizer/PostEscapeAnalysis.cpp]: Fixed in commit https://github.com/eclipse/openj9/pull/6654/commits/7f10d28781bc910fa1fcf09c5a429a2be1e55cf1

**Comment by vijaysun-omr** [comment]
Thanks @hzongaro I have approved this. @andrewcraik please review as well and merge if you are good with it.

File: runtime/compiler/optimizer/PreEscapeAnalysis.hpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      }
+
+   virtual int32_t perform();
+   virtual const char * optDetailString() const throw();
+
+   private:
+   NodeDeque *_loads;
+   void insertFakePrepareForOSR(TR::Block *block, TR::Node *induceCall);
+   void processAutosAndPendingPushes(TR::ResolvedMethodSymbol *rms, TR_OSRMethodData *methodData, int32_t byteCodeIndex);
+   void processSymbolReferences(TR_Array<List<TR::SymbolReference>> *symbolReferences, TR_BitVector *deadSymRefs);
```
Comment (andrewcraik) [inline runtime/compiler/optimizer/PreEscapeAnalysis.hpp]: Can you please doxygen these 3 APIs?

Comment (hzongaro) [reply runtime/compiler/optimizer/PreEscapeAnalysis.hpp]: Thanks, Andrew @andrewcraik .  Added in https://github.com/eclipse/openj9/pull/6654/commits/5b375e5237119cdd8cd2ac8c737e279802c146d8.
If you're OK with that change, I will squash that commit as well as https://github.com/eclipse/openj9/pull/6654/commits/7f10d28781bc910fa1fcf09c5a429a2be1e55cf1

Comment (andrewcraik) [reply runtime/compiler/optimizer/PreEscapeAnalysis.hpp]: @hzongaro doxygen looks good - please squash

Comment (hzongaro) [reply runtime/compiler/optimizer/PreEscapeAnalysis.hpp]: @andrewcraik Squashed.

**Comment by andrewcraik** [comment]
Jenkins test sanity xlinux,win,plinux jdk8,jdk11

**Pull request merged**